src/formats/jpeg/writer.zig

```diff
@@ -0,0 +1,849 @@
+            0xe2, 0xe3, 0xe4, 0xe5, 0xe6, 0xe7, 0xe8, 0xe9,
+            0xea, 0xf2, 0xf3, 0xf4, 0xf5, 0xf6, 0xf7, 0xf8,
+            0xf9, 0xfa,
+        },
+    },
+};
+
+// Unscaled quantization tables in zig-zag order. Each encoder copies and scales
+// the tables according to its quality parameter.
+// The values are derived from section K.1 of the spec, after converting from
+// natural to zig-zag order.
+const unscaledQuant = [quant_index][block_size]u8{
+    // Luminance.
+    [_]u8{
+        16,  11,  12,  14,  12,  10,  16,  14,
+        13,  14,  18,  17,  16,  19,  24,  40,
+        26,  24,  22,  22,  24,  49,  35,  37,
+        29,  40,  58,  51,  61,  60,  57,  51,
+        56,  55,  64,  72,  92,  78,  64,  68,
+        87,  69,  55,  56,  80,  109, 81,  87,
+        95,  98,  103, 104, 103, 62,  77,  113,
+        121, 112, 100, 120, 92,  101, 103, 99,
+    },
+    // Chrominance.
+    [_]u8{
+        17, 18, 18, 24, 21, 24, 47, 26,
+        26, 47, 99, 66, 56, 66, 99, 99,
+        99, 99, 99, 99, 99, 99, 99, 99,
+        99, 99, 99, 99, 99, 99, 99, 99,
+        99, 99, 99, 99, 99, 99, 99, 99,
+        99, 99, 99, 99, 99, 99, 99, 99,
+        99, 99, 99, 99, 99, 99, 99, 99,
+        99, 99, 99, 99, 99, 99, 99, 99,
+    },
+};
+
+// Huffman LUT entry: 8-bit size in MSB, 24-bit code in LSB
+const default_huffman_lut = blk: {
+    var tables: [nHuffIndex][256]u32 = undefined;
+    for (theHuffmanSpec, 0..) |spec, idx| {
+        tables[idx] = initHuffmanLUT(spec);
+    }
+    break :blk tables;
+};
+
+// Zigzag ordering for DCT coefficients (same as JPEG spec)
+const unzig = [block_size]u8{
+    0,  1,  8,  16, 9,  2,  3,  10,
+    17, 24, 32, 25, 18, 11, 4,  5,
+    12, 19, 26, 33, 40, 48, 41, 34,
+    27, 20, 13, 6,  7,  14, 21, 28,
+    35, 42, 49, 56, 57, 50, 43, 36,
+    29, 22, 15, 23, 30, 37, 44, 51,
+    58, 59, 52, 45, 38, 31, 39, 46,
+    53, 60, 61, 54, 47, 55, 62, 63,
+};
+
+// div returns a/b rounded to the nearest integer, instead of rounded to zero
+fn div(a: i32, b: i32) i32 {
+    if (a >= 0) {
+        return @divTrunc((a + (b >> 1)), b);
+    } else {
+        return -@divTrunc((-a + (b >> 1)), b);
+    }
+}
+
+// Initialize Huffman LUT from Huffman specification
+fn initHuffmanLUT(spec: HuffmanSpec) [256]u32 {
+    var lut: [256]u32 = [_]u32{0} ** 256;
+
+    var code: u32 = 0;
+    var value_index: usize = 0;
+
+    for (spec.count, 0..) |count, depth| {
+        const bit_count = @as(u32, @intCast(depth + 1));
+        for (0..count) |_| {
+            std.debug.assert(value_index < spec.value.len);
+            const symbol = spec.value[value_index];
+            lut[symbol] = (bit_count << 24) | code;
+            code += 1;
+            value_index += 1;
+        }
+        code <<= 1;
+    }
+
+    return lut;
+}
+
+/// JPEGWriter handles encoding images to JPEG format
+pub const JPEGWriter = struct {
+    writer: *std.Io.Writer,
+    err: ?Image.WriteError = null,
```

Comment:
I fail to see why you need to store an error and do early exits when the `try` keyword already does this for you.